packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,191 @@
+import { IntervalCollection, SequenceInterval } from "./intervalCollection";
+import { SharedString } from "./sharedString";
+
+const IntervalEventType = {
+	CHANGE: 0,
+	ADD: 1,
+	DELETE: 2,
+	PROPERTYCHANGED: 3,
+} as const;
+
+type IntervalEventType = typeof IntervalEventType[keyof typeof IntervalEventType];
+
+type IntervalRevertible =
+	| {
+			event: typeof IntervalEventType.CHANGE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.ADD;
+			interval: SequenceInterval;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.DELETE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.PROPERTYCHANGED;
+			interval: SequenceInterval;
+			propertyDeltas: PropertySet;
+	  };
+
+type TypedRevertible<T extends IntervalRevertible["event"]> = IntervalRevertible & { event: T };
+
+export function appendLocalAddToRevertibles(
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	revertibles.push({
+		event: IntervalEventType.ADD,
+		interval,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalDeleteToRevertibles(
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
+		end: interval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of previous interval
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: previousInterval.start,
+		end: previousInterval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
```

Comment:
I think it's true but worth checking with Abram that an interval id is sufficiently reliable across edits to guarantee that it is the interval we think it is.